monitor/event_stats.hpp

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2015+ Budnik Andrey <budnik27@gmail.com>
+ *
+ * This file is part of Elliptics.
+ *
+ * Elliptics is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Elliptics is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Elliptics.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef EVENT_STATS_HPP
+#define EVENT_STATS_HPP
+
+#include "cache/treap.hpp"
+
+#include <algorithm>
+#include <functional> // not2
+#include <stack>
+
+//#define TOP_SLICES
+#define TOP_LRU
+
+#ifdef TOP_LRU
+
+namespace ioremap { namespace monitor {
+
+template<typename E>
+class key_stat_t;
+
+}}  /* namespace ioremap::monitor */
+
+namespace ioremap { namespace cache {
+
+template<typename E>
+struct treap_node_traits<ioremap::monitor::key_stat_t<E> >
+{
+	typedef typename E::key_type key_type;
+	typedef typename E::time_type priority_type;
+};
+
+}}  /* namespace ioremap::cache */
+
+namespace ioremap { namespace monitor {
+
+class mutex_lock_policy
+{
+	typedef std::mutex mutex_type;
+public:
+	class unique_lock {
+	public:
+		unique_lock(mutex_lock_policy *policy)
+		: lock_(policy->get_lock())
+		{}
+	private:
+		std::unique_lock<mutex_type> lock_;
+	};
+
+	mutex_type &get_lock() { return mut_; }
+private:
+	mutex_type mut_;
+};
+
+class null_lock_policy
+{
+public:
+	class unique_lock {
+	public:
+		unique_lock(null_lock_policy *policy) { (void) policy; }
+	};
+};
+
+template<typename E>
+class key_stat_t : public ioremap::cache::treap_node_t<key_stat_t<E> > {
```

Comment:
Why did you make two classes key_stat_event and key_stat_t? Please do not make extra templates and meta-classes where they aren't needed.